src/audio/volume/volume.c

```diff
@@ -415,6 +418,59 @@ static int volume_free(struct processing_module *mod)
 	return 0;
 }
 
+static void set_linear_ramp_coef(struct vol_data *cd, int chan, bool constant_rate_ramp)
+{
+	int32_t delta;
+	int32_t delta_abs;
+	int32_t coef;
+
+	if (!cd->initial_ramp)
+		return;
+
+	/* Get volume transition delta and absolute value */
+	delta = cd->tvolume[chan] - cd->volume[chan];
+	if (!delta) {
+		cd->ramp_coef[chan] = 0;
+		return;
+	}
+
+	delta_abs = ABS(delta);
+
+	/* The ramp length (initial_ramp [ms]) describes time of mute
+	 * to vol_max unmuting. Normally the volume ramp has a
+	 * constant linear slope defined this way and variable
+	 * completion time. However in streaming start it is feasible
+	 * to apply the entire topology defined ramp time to unmute to
+	 * any used volume. In this case the ramp rate is not constant.
+	 * Note also the legacy mode without known vol_ramp_range where
+	 * the volume transition always uses the topology defined time.
+	 */
+	if (constant_rate_ramp && cd->vol_ramp_range > 0)
+		coef = cd->vol_ramp_range;
+	else
+		coef = delta_abs;
+
+	/* Divide and round to nearest. Note that there will
+	 * be some accumulated error in ramp time the longer
+	 * the ramp and the smaller the transition is.
+	 */
+	coef = (2 * coef / cd->initial_ramp + 1) >> 1;
```

Comment:
It's rounded division.

Reply:
@singalsu no, my proposal has exactly the same number of divisions as yours - one division and one shift to the right. I understand that it's rounding to the nearest, but as I said, I'm more used to these three forms of rounding down, nearest and up:
```
a / b
(a + b / 2) / b
(a + b - 1) / b
```
is yours `(2 * a / b + 1) / 2` also a common form? I'm just not used to it, so was surprised

Reply:
I wonder if your proposal is losing precision because with b / 2 integer divide. For sure my 2 * a is not losing precision but limits the max value of a more (this constraint has been checked). If you are not OK with this press red and demonstrate with a practical example how your version is more accurate or better, if this OK press green. Also I didn't change this equation in this patch.

Reply:
Anyway, in all of these there's the idea to make an intermediate calculation result that is 2 * result, add 1 (half of final LSB or 0.5 if thinking as float or fraction it), finally truncate the unnecessary bit or cancel the 2 * result scaling. Nothing new invented.

Reply:
> I wonder if your proposal is losing precision because with b / 2 integer divide. For sure my 2 * a is not losing precision but limits the max value of a more (this constraint has been checked). If you are not OK with this press red and demonstrate with a practical example how your version is more accurate or better, if this OK press green. Also I didn't change this equation in this patch.

@singalsu sorry, no, I wasn't saying that mine was better! I just said, that it was something I was used to and yours was new to me, I needed some time to understand it, so I was just *academically* curious whether it's an up-to-now unknown to me common solution or something that you came up with. I'm perfectly fine with your solution, but when I next time read it in the code, I suspect I'll stumble and pause again :-)

Reply:
> I wonder if your proposal is losing precision because with b / 2 integer divide. For sure my 2 * a is not losing precision but limits the max value of a more (this constraint has been checked). If you are not OK with this press red and demonstrate with a practical example how your version is more accurate or better, if this OK press green. Also I didn't change this equation in this patch.

@singalsu sorry, no, I wasn't saying that mine was better! I just said, that it was something I was used to and yours was new to me, I needed some time to understand it, so I was just *academically* curious whether it's an up-to-now unknown to me common solution or something that you came up with. I'm perfectly fine with your solution, but when I next time read it in the code, I suspect I'll stumble and pause again :-)